analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsStateless.cs

```diff
@@ -0,0 +1,96 @@
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    public static class Fixme   // FIXME: Remove temporary workaround
+    {
+        public static bool IsAzureFunction(this SyntaxNodeAnalysisContext context) =>
+            context.ContainingSymbol.GetAttributes().Any(x => x.AttributeClass.Name.Contains("FunctionName"));
+    }
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class AzureFunctionsStateless : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S6419";
+        private const string MessageFormat = "Do not modify a static state from Azure Function.";
+
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c => CheckTarget(c, ((AssignmentExpressionSyntax)c.Node).Left),
+                SyntaxKind.SimpleAssignmentExpression,
+                SyntaxKind.AddAssignmentExpression,
+                SyntaxKind.SubtractAssignmentExpression,
+                SyntaxKind.MultiplyAssignmentExpression,
+                SyntaxKind.DivideAssignmentExpression,
+                SyntaxKind.ModuloAssignmentExpression,
+                SyntaxKind.AndAssignmentExpression,
+                SyntaxKind.ExclusiveOrAssignmentExpression,
+                SyntaxKind.OrAssignmentExpression,
+                SyntaxKind.LeftShiftAssignmentExpression,
+                SyntaxKind.RightShiftAssignmentExpression,
+                SyntaxKindEx.CoalesceAssignmentExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c => CheckTarget(c, ((PrefixUnaryExpressionSyntax)c.Node).Operand),
+                SyntaxKind.PreDecrementExpression,
+                SyntaxKind.PreIncrementExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c => CheckTarget(c, ((PostfixUnaryExpressionSyntax)c.Node).Operand),
+                SyntaxKind.PostDecrementExpression,
+                SyntaxKind.PostIncrementExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var argument = (ArgumentSyntax)c.Node;
+                    if (argument.RefOrOutKeyword != default)
+                    {
+                        CheckTarget(c, argument.Expression);
+                    }
+                },
+                SyntaxKind.Argument);
+        }
+
+        private static void CheckTarget(SyntaxNodeAnalysisContext context, ExpressionSyntax target)
+        {
+            if (context.IsAzureFunction()
+                && context.SemanticModel.GetSymbolInfo((target as ElementAccessExpressionSyntax)?.Expression ?? target).Symbol is { } symbol
```

Comment:
I would prefer to move the `(target as ElementAccessExpressionSyntax)?.Expression ?? target` target part to the caller. The caller already extracts the right expression from the node and should do so entirely if possible.